docs/design/features/OnStackReplacement.md

```diff
@@ -0,0 +1,1095 @@
+this set only the values it sees as truly live.
+
+This means that we can run optimizations in the original method so long as they
+do not alter the computation of the over-approximated live set at any
+patchpoint.
+
+The proposed Tier0 optimizations fit into this category, so long as we restrict
+patchpoints to stack-empty points: we may prune away unreachable code paths (say
+from HW intrinsic checks or provably true or false predicate evaluations &mdash
+;patchpoints in pruned sections would be unreachable) and simplify computations.
+Optimizing expressions may reduce the truly live set but so long as all stores
+to locals and args are kept live the base values needed for any alternate
+version of the code will be available.
+
+### 4.7 Alternative Method Optimizations
+
+In options where the alternative method has multiple entry points, one must be
+wary of early aggressive optimizations done when optimizing the alternative. The
+original version of the method may hit a patchpoint while executing code that
+can be optimized away by the more aggressive alternative method compiler (e.g.
+it may be executing a series of type equality tests in a generic method that the
+optimizing jit can evaluate at jit time). But with our simple patchpoint
+recognition algorithm the alternate compiler can quickly verify that the
+patchpoint IL offset is a viable entry point and ensure that the code at that
+offset is not optimized away. If it turns out that the entry point code is
+optimizable then we may choose to peel one iteration from the entry point loop
+(because with our patchpoint strategy, execution in the alternate method will
+immediately hit a loop top once it is out of the prolog) and allow the in-loop
+versions to be optimized.
+
+###  4.8 Prologs and Unwind
+
+The alternative version of the method will, in all likelihood, need to save and
+restore a different set of callee-saves registers than the original version. But
+since the original stack frame has already saved some registers, the alternative
+version prolog will either need to save a superset of those registers or else
+restore the value of some registers in its prolog. So, the alternative version
+needs to know which registers the original saved and where in the stack they are
+stored.
+
+If we want to preserve frame offsets for address-taken locals then we may face a
+conflict as altering the number of callee save slots may alter frame offsets for
+locals. One thought here is that we could perhaps implement a chained unwind
+scheme, where there is an initial prolog that emulates the original version
+prolog and duplicates its saves, and then a subsequent "shrink wrapped" prolog
+&amp; epilog that saves any additional registers in a disjoint area.
+
+**Prototype: runtime recovers most callee-saves from original method via virtual
+unwind; new method just saves what is needed in its own save area. The only
+register save live on the old frame is the saved FP.**
+
```

Comment:
When it is time to transition, the patchpoint helper virtually unwinds two frames from the stack -- its own frame, and the frame for the original method. So the unwound context restores the callee saves done by the original method. That turns out to be sufficient.

You might think we would need to save all the register state when we call the patchpoint helper, but that's not the case. Because the original method is un-optimized, there isn't any live IL state in registers across the call to the patchpoint helper -- all the live IL state for the method is on the original frame -- so the argument and caller-save registers are dead at the patchpoint. Thus only part of register state that is significant for ongoing computation is the callee-saves, and these we recover via unwind.

We then "call" the OSR method from this unwound context. It does its own callee-saves as needed, and recovers the arguments/IL state from the original frame.

If we were to support patchpoints in optimized code things would be more complicated.

Reply:
Thanks - perhaps you could elaborate on this just a bit further in the document.

Reply:
Done.